Arcs join support

skef opened the pull request:
This PR adds support for Arcs joins to the Expand Stroke facility. There are also some associated UI improvements and some (unfiled) bug fixes. 

See https://svgwg.org/svg2-draft/painting.html#LineJoin and also sections 13.5.7 through 9 for more information about Arcs joins (relative to the included FontForge doc). http://tavmjong.free.fr/SVG/LINEJOIN_STUDY/ may also be of interest. 

As is probably evident from the code this was a bit of a pain to implement; probably more of a pain than the SVG 2 authors yet appreciate. (However, as with the other joins and caps part of the pain comes from generalizing the arcs join pattern for arbitrary nibs. Unclipped they generalize without modification but as the docs note the SVG 2 arcs clipping heuristic fails for some nib/limit combinations.)

Still, this join style is a useful and significant addition. In many cases it produces better output when increasing the weight of a glyph with "Change Weight ...". I therefore plan to add it as an alternative join option for that facility soon. (Unfortunately it does not *always* produce better output. If you're curious you can test Expand Stroke... on a given glyph using the "External Only" option.)

ctrlcctrlv commented:
I'm sorry I didn't get the chance to test this sooner. Guess I know what I'll be doing tomorrow :-)

skef commented:
@ctrlcctrlv I only pointed you at the previous version of the branch in case you were interested; it wasn't quite working then anyway. I only finished the Ratio algorithm yesterday and made the docs and (minimal) automated test changes today.

ctrlcctrlv commented:
I'm sorry @skef&mdash;the branch compiles fine for me, but the &laquo;Arcs&raquo; join type behaves exactly the same as the &laquo;Miter&raquo; join type. I also get a lot of:

```
SplinePointListIsClockwise: Found error
```

(I'm on IRC if you want.)

skef commented:
An arcs join should have the same effect when the source contours are straight (have no curvature) at the join, so whether this is a problem depends on specific shape you tested with. 

The first diagram in https://svgwg.org/svg2-draft/painting.html#LineJoin illustrates what the difference should be.

skef commented:
Well, I spent quite a bit of yesterday and today on what I thought would be a better overall clipping algorithm, so that someone who didn't care could just leave that one selected. Unfortunately this alternate approach fails in other cases I didn't anticipate, so for now the current split between SVG 2 and Ratio will have to do. (Ratio does have the distinct advantage of conceptual simplicity, so that outright failure cases are unlikely.)

skef commented:
@jtanx if you have time to give this a look like you did for the main putback I would appreciate it. Even though some of that code was too obscure to dive into you still found a pile of problems.

skef commented on fontforge/splinestroke.c:
```diff
@@ -919,19 +1029,38 @@ static void SSAppendArc(SplineSet *cur, bigreal major, bigreal minor,
     _CalcNibOffset(nc, n, ut_fm, false, &no_fm, -1);
     _CalcNibOffset(nc, n, ut_to, false, &no_to, -1);
+    if ( limit ) {
+	if (    !bk && no_fm.nci[false] == no_to.nci[false]
```
That's a 2 element array that corresponds to an "is_ccw" condition, where I'm picking one value arbitrarily because I know the two values won't be different in this case. I'm using "false" because of that pattern but I'm happy to just use 0 instead.

skef commented on fontforge/utanvec.h:
```diff
@@ -36,12 +36,15 @@
 #define BP_REV(v) (BasePoint) { -(v).x, -(v).y }
```
I'd be thrilled to switch to inline functions. I didn't know they were "allowed". (Are you *sure* they're "allowed"? I haven't run into any in other sources and some people don't like/trust them.)

jtanx replied:
Oops, I guess I was thinking more about how it worked under C++.

I personally don't have anything against them, and I think it reads easier, while also preventing multiple evaluations of parameters, unlike the macros.

Might need to stick a `static` in front of the inline declarations I suggested, to prevent linkage issues ([[1](https://stackoverflow.com/questions/34139737/inline-function-in-header-file-in-c)], [[2](https://www.greenend.org.uk/rjk/tech/inline.html)]).

skef replied:
Yeah, it does seem like static inline is the way to go for these, if we can. 

As I indicated I am very happy to use C99 static inline functions in preference to macros. @frank-trampe would that be OK? Can we add that as an optional FontForge semantic?

ctrlcctrlv replied:
I already thought it was allowed because we previously agreed all C99 features are allowed, so I support

jtanx replied:
Just clarifying my stance a little, I'm ok with inline functions in headers to be used judiciously; i.e. where it makes sense / where the function is trivially inlinable. I wouldn't be ok with trying to do this with more complex functions. 

So basically use common sense ;)

ctrlcctrlv replied:
Of course, no code is ever guaranteed to be accepted, and just because something is allowed doesn't mean it will be accepted :-)

skef replied:
I primarily have a "when in Rome" approach to projects. I certainly don't indent my own code the way `spinestroke.c` is indented. 

The topic of discussion here, as I see it, is not as much how to avoid function-call overhead (which is relatively small these days) but how to symbolically represent a or chain of computations that you still want to be locally optimized.

GW's approach, it seems, was to do it in his head and write everything out every time. That's ... not ideal practice. I've been using macros because those were the standard approach at the time people were writing GW-like code. As @jtanx implies, the general problem with macros is that the compiler loses the syntax tree and the optimizer can't always put it back together. 

Inline functions -- and in C99 in particular *static* inline functions -- are the better way of handling the problem. The compiler gets the tree and therefore can locally optimize everything with much better information. With the "static" it has the option of bailing out by adding a local function. (Naively one might think that if it's non-inlined once it will never be inlined, but that actually depends on whether there are any constant/compile-time-determinable arguments.)

jtanx commented:
I get the following warnings when turning up the warning level:

```
../fontforge/splinestroke.c: In function 'ArcClip':
../fontforge/splinestroke.c:1493:18: warning: unused variable 'ci' [-Wunused-variable]
 1493 |     BasePoint v, ci;
      |                  ^~
../fontforge/splinestroke.c: In function 'RoundJoin':
../fontforge/splinestroke.c:1675:10: warning: unused variable 'trans' [-Wunused-variable]
 1675 |     real trans[6];
      |          ^~~~~
../fontforge/splinestroke.c: In function 'ArcsJoin':
../fontforge/splinestroke.c:1856:9: warning: using integer absolute value function 'abs' when argument is of floating point type 'double' [-Wabsolute-value]
 1856 |  r_fm = abs(1.0/K_fm);
      |         ^~~
../fontforge/splinestroke.c:1861:9: warning: using integer absolute value function 'abs' when argument is of floating point type 'double' [-Wabsolute-value]
 1861 |  r_to = abs(1.0/K_to);
      |         ^~~
../fontforge/splinestroke.c:1775:13: warning: variable 'jlen' set but not used [-Wunused-but-set-variable]
 1775 |     bigreal jlen, jlim, r_clip, clip_ratio, start_angle, end_angle;
      |             ^~~~
../fontforge/splinestroke.c: In function 'HandleCap':
../fontforge/splinestroke.c:2125:57: warning: unused variable 'p2' [-Wunused-variable]
 2125 |     BasePoint refp = BP_ADD(sxy, c->pseudo_origin), p1, p2;
      |                                                         ^~
../fontforge/splinestroke.c:2125:53: warning: unused variable 'p1' [-Wunused-variable]
 2125 |     BasePoint refp = BP_ADD(sxy, c->pseudo_origin), p1, p2;
      |                                                     ^~
../fontforge/splinestroke.c:2125:15: warning: unused variable 'refp' [-Wunused-variable]
 2125 |     BasePoint refp = BP_ADD(sxy, c->pseudo_origin), p1, p2;
      |               ^~~~

```

jtanx commented on fontforgeexe/cvstroke.c:
```diff
@@ -1170,6 +1296,8 @@ static void MakeStrokeDlg(void *cv, void (*strokeit)(void *,StrokeInfo *,int),
 
     StrokeSetup(&sd,si->stroke_type);
+    JoinSetup(&sd,si->join==lj_arcs?2:((si->join==lj_miterclip||si->join==lj_miter)?1:0));
+    ExtendSetup(&sd,si->join==lc_butt||si->join==lc_round);
```
```
../fontforgeexe/cvstroke.c: In function 'MakeStrokeDlg':
../fontforgeexe/cvstroke.c:1300:29: warning: comparison between 'enum linejoin' and 'enum linecap' [-Wenum-compare]
 1300 |     ExtendSetup(&sd,si->join==lc_butt||si->join==lc_round);
      |                             ^~
../fontforgeexe/cvstroke.c:1300:48: warning: comparison between 'enum linejoin' and 'enum linecap' [-Wenum-compare]
 1300 |     ExtendSetup(&sd,si->join==lc_butt||si->join==lc_round);
```

skef replied:
Oh, yeah, that's an outright bug.

skef commented on fontforgeexe/cvstroke.c:
```diff
@@ -1093,11 +1219,11 @@ static void MakeStrokeDlg(void *cv, void (*strokeit)(void *,StrokeInfo *,int),
-	checkarray[1][1] = &gcd[gcdoff-1]; checkarray[1][2] = GCD_Glue;
-	checkarray[1][3] = checkarray[2][0] = NULL;
+	checkarray[1] = &gcd[gcdoff-1]; checkarray[2] = GCD_Glue;
+	checkarray[2] = NULL;
```
It's probably not OK to use GHVBoxCreate (as opposed to GHBoxCreate) for a one-row structure, or at least not without faking the arrays appropriately. I'll fix the mislabeled one and try to verify.

skef commented:
I rebased to pick up the sanitizer changes, and therefore force-pushed. @jtanx I believe I've addressed everything you found (those `abs()` calls were a bad, but I guess not too harmful at normal scales... )

I did go ahead and change the macros to inlines, but in a separate commit. This way we can consider the pluses and minuses with working code.

jtanx commented on fontforge/splinestroke.c:
```diff
@@ -1518,16 +1778,304 @@ static void MiterJoin(JoinParams *jpp) {
+    neg_fm = (K_fm < 0);
+    neg_to = (K_to < 0);
+
+    if ( K_fm!=0.0 ) {
```
Logic seems ok, but it's still raising this (probably false) warning:

```
../fontforge/splinestroke.c: In function 'HandleJoin':
../fontforge/splinestroke.c:1878:6: warning: 'r_to' may be used uninitialized in this function [-Wmaybe-uninitialized]
 1878 |      IntersectLineCircle(p_fm, jpp->ut_fm, center_to, r_to, &i, &i2);
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../fontforge/splinestroke.c:1782:59: note: 'r_to' was declared here
 1782 |     bigreal t, K_fm, K_to, fsw_fm, fsw_to, fsw_avg, r_fm, r_to, i_dist;
      |                                                           ^~~~
../fontforge/splinestroke.c:2024:14: warning: 'r_fm' may be used uninitialized in this function [-Wmaybe-uninitialized]
 2024 |      ci_fm = ArcClip(center_fm, r_fm, neg_fm, p_fm, i, clip_ratio);
      |              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../fontforge/splinestroke.c:1782:53: note: 'r_fm' was declared here
 1782 |     bigreal t, K_fm, K_to, fsw_fm, fsw_to, fsw_avg, r_fm, r_to, i_dist;
      |                                                     ^~~~
```

It might go away if using explicit booleans for K_fm==0 / K_to == 0?

ctrlcctrlv commented:
This isn't related to the patch, but I just want to say it:

We are very, very happy to have you back @jtanx. I almost cut a release without you (#3933), but I have no idea how to build on Mac and don't have a Mac, so was apprehensive to not test at all on Mac OS as so many of our users have it.

Had you not returned for another month, however, I would have pulled the trigger. So I'm really glad to have you back, and I hope you can help us cut a release soon. You've taught me a lot and your presence improves every PR.

jtanx commented:
Yeah, I've been around, just not very active; I was trying to focus on some other projects (still nowhere near done btw). Anyway here's me back procrastinating on FF instead...

skef commented:
OK I pushed a commit with the boolean trick. Seems to work around the issue but I may not be verifying right.

skef commented:
It's only fair -- the warning flag has "maybe" in it, after all.

skef commented:
We never got consensus on the change to inline functions, but we asked and it seems to work well. So I'm going to merge this.

frank-trampe commented:
For what it's worth, I had no objection to the inline functions. Apparently, nobody else did either.

skef commented:
@frank-trampe I think that's worth quite a bit -- I'd say that's a rough consensus among current regular/general contributors.
